firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,296 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
 
-                (found_key, value)
+                if !next_child_success {
+                    return Ok(None);
+                }
             }
-            _ => unreachable!(),
-        };
+        }
+
+        first_loop = false;
+    }
+}
+
+fn get_children_iter(branch: &BranchNode) -> impl Iterator<Item = (DiskAddress, u8)> {
+    branch
+        .children
+        .into_iter()
+        .enumerate()
+        .filter_map(|(pos, child_addr)| child_addr.map(|child_addr| (child_addr, pos as u8)))
+}
+
+#[must_use]
+struct MustUse<T>(T);
+
+impl<T> From<T> for MustUse<T> {
+    fn from(t: T) -> Self {
+        Self(t)
+    }
+}
 
-        Poll::Ready(Some(Ok(return_value)))
+impl<T: Not> Not for MustUse<T> {
+    type Output = T::Output;
+
+    fn not(self) -> Self::Output {
+        self.0.not()
     }
 }
 
-/// Compute a key from a set of parents
-fn key_from_parents(parents: &[(ObjRef, u8)]) -> Vec<u8> {
-    parents[1..]
-        .chunks_exact(2)
-        .map(|parents| (parents[0].1 << 4) + parents[1].1)
-        .collect::<Vec<u8>>()
+fn next_child<'a, S, T, Iter>(
+    merkle: &'a Merkle<S, T>,
+    mut children: Iter,
+    parents: &mut Vec<(ObjRef<'a>, u8)>,
+    node: &mut ParentNode<'a>,
+    pos: &mut u8,
+) -> Result<MustUse<bool>, super::MerkleError>
+where
+    Iter: Iterator<Item = (DiskAddress, u8)>,
+    S: ShaleStore<Node>,
+{
+    if let Some((child_addr, child_pos)) = children.next() {
+        let child = merkle.get_node(child_addr)?;
+
+        *pos = child_pos;
+        let node = std::mem::replace(node, ParentNode::New(child));
+        parents.push((node.into_node(), *pos));
+    } else {
+        let Some((next_parent, next_pos)) = parents.pop() else {
+            return Ok(false.into());
+        };
+
+        *node = ParentNode::Visited(next_parent);
+        *pos = next_pos;
+    }
+
+    Ok(true.into())
 }
-fn key_from_parents_and_leaf(parents: &[(ObjRef, u8)], leaf: &LeafNode) -> Vec<u8> {
-    let mut iter = parents[1..]
+
+/// create an iterator over the key-nibbles from all parents _excluding_ the sentinal node.
+fn nibble_iter_from_parents<'a>(parents: &'a [(ObjRef, u8)]) -> impl Iterator<Item = u8> + 'a {
+    parents
         .iter()
-        .map(|parent| parent.1)
-        .chain(leaf.path.to_vec());
-    let mut data = Vec::with_capacity(iter.size_hint().0);
-    while let (Some(hi), Some(lo)) = (iter.next(), iter.next()) {
+        .skip(1) // always skip the sentinal node
+        .flat_map(|(parent, child_nibble)| match parent.inner() {
+            NodeType::Branch(_) => vec![*child_nibble],
+            NodeType::Extension(extension) => extension.path.to_vec(),
+            NodeType::Leaf(leaf) => leaf.path.to_vec(),
+        })
+}
+
+fn key_from_nibble_iter<Iter: Iterator<Item = u8>>(mut nibbles: Iter) -> Vec<u8> {
+    let mut data = Vec::with_capacity(nibbles.size_hint().0);
```

Comment:
Not going to change this right away, but noticed reviewing my own code

```suggestion
    let mut data = Vec::with_capacity(nibbles.size_hint().0 / 2);
```